### ISSUE TITLE
Cannot build GEOSX on Pecan with GPU support

### ISSUE THREAD

**Issue by TotoGaz** [GEOSX/GEOSX#1296]
Compilation fails when building GEOSX (version https://github.com/GEOSX/GEOSX/commit/95da07757535c28a6e557012ef677f006e1f6cdb) with TPLs (version https://github.com/GEOSX/thirdPartyLibs/commit/0968c98637153c750f0916e3dafabd8f9f630ced, built with GPU support)

I use gcc `8.2.0`, nvcc `release 10.1, V10.1.168`.

CUDA flags are `-restrict -arch sm_75 --expt-relaxed-constexpr --expt-extended-lambda -Werror cross-execution-space-call,reorder,deprecated-declarations -Xcompiler -std=c++14`

Error message is
```
/data/gpfs/Users/j0436735/GEOSX/GEOSX/src/coreComponents/LvArray/src/arrayManipulation.hpp(153): error: calling a __host__ function("geosx::EpetraMatrix::~EpetraMatrix [subobject]") from a __device__ function("destroy") is not allowed
          detected during:
            instantiation of "void LvArray::arrayManipulation::destroy(T *, std::ptrdiff_t) [with T=geosx::TrilinosInterface::ParallelMatrix]"
/data/gpfs/Users/j0436735/GEOSX/GEOSX/src/coreComponents/LvArray/src/bufferManipulation.hpp(191): here
            instantiation of "void LvArray::bufferManipulation::free(BUFFER &, std::ptrdiff_t) [with BUFFER=LvArray::ChaiBuffer<geosx::TrilinosInterface::ParallelMatrix>]"
/data/gpfs/Users/j0436735/GEOSX/GEOSX/src/coreComponents/LvArray/src/Array.hpp(146): here
            instantiation of "LvArray::Array<T, NDIM, PERMUTATION, INDEX_TYPE, BUFFER_TYPE>::~Array() [with T=geosx::TrilinosInterface::ParallelMatrix, NDIM=2, PERMUTATION=camp::make_idx_seq_t<2L>, INDEX_TYPE=geosx::localIndex, BUFFER_TYPE=LvArray::ChaiBuffer]"
/data/gpfs/Users/j0436735/GEOSX/GEOSX/src/coreComponents/linearAlgebra/utilities/BlockOperator.hpp(69): here

/data/gpfs/Users/j0436735/GEOSX/GEOSX/src/coreComponents/LvArray/src/arrayManipulation.hpp(153): error: calling a __host__ function("geosx::HypreMatrix::~HypreMatrix [subobject]") from a __device__ function("destroy") is not allowed
          detected during:
            instantiation of "void LvArray::arrayManipulation::destroy(T *, std::ptrdiff_t) [with T=geosx::HypreInterface::ParallelMatrix]"
/data/gpfs/Users/j0436735/GEOSX/GEOSX/src/coreComponents/LvArray/src/bufferManipulation.hpp(191): here
            instantiation of "void LvArray::bufferManipulation::free(BUFFER &, std::ptrdiff_t) [with BUFFER=LvArray::ChaiBuffer<geosx::HypreInterface::ParallelMatrix>]"
/data/gpfs/Users/j0436735/GEOSX/GEOSX/src/coreComponents/LvArray/src/Array.hpp(146): here
            instantiation of "LvArray::Array<T, NDIM, PERMUTATION, INDEX_TYPE, BUFFER_TYPE>::~Array() [with T=geosx::HypreInterface::ParallelMatrix, NDIM=2, PERMUTATION=camp::make_idx_seq_t<2L>, INDEX_TYPE=geosx::localIndex, BUFFER_TYPE=LvArray::ChaiBuffer]"
/data/gpfs/Users/j0436735/GEOSX/GEOSX/src/coreComponents/linearAlgebra/utilities/BlockOperator.hpp(69): here

2 errors detected in the compilation of "/tmp/tmpxft_00014dd8_00000000-6_BlockPreconditioner.cpp1.ii".
```

**Comment by klevzoff** [comment]
This error does not appear with CUDA toolkit `10.1.243` and higher. Previous versions seem to randomly ignore the pragma that is supposed to silence this.

In any case, the only reason to use `LvArray` containers there was the convenience of 2D-indexing. They can be replaced with `std::vector` if need be.

**Comment by TotoGaz** [comment]
Fixed with a `10.2` version of CUDA.
